In [1]:
#imports
import pandas as pd
%matplotlib inline 
import matplotlib.pyplot as plt
from matplotlib import dates as mpl_dates
from matplotlib.pyplot import figure
import numpy as np
import seaborn as sns
from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()
import calendar

In [2]:
df_merchandise = pd.read_excel('Merchandise.xlsx', index_col="ProductID")

## Data Preprocessing & Analysis

In [3]:
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 0             filler       filler      filler                    filler   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 0               NaN    0.00 1999-12-31   2018-04-14          68.00   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   

          TransferInd  ... ListOnline  CustNo  FixSplit  TaxRate  PhyDate  \
ProductID              ...                                                  
-2               Sold  ...      False     NaN     False      NaN      NaN   
-1               Sold  ...      False     NaN     False      NaN      NaN   
 0               Sold  ...      False     NaN     False      NaN      NaN   
 1               Sold  ...      False     NaN     False      NaN      NaN   
 2               Sold  ...      False     NaN     False      NaN      NaN   

          Auction2 ConCode BuyOut  CardFee  Reminded  
ProductID                                             
-2             NaN     NaN  False      0.0     False  
-1             NaN     NaN  False      0.0     False  
 0             NaN     NaN  False      0.0     False  
 1             NaN     NaN  False      0.0     False  
 2             NaN     NaN  False      0.0     False  

[5 rows x 50 columns]

In [4]:
df_merchandise.shape

(15308, 50)

In [5]:
df_merchandise.count()

ProductType           15308
ProductColor          15308
ProductSize           15308
ProductDescription    15308
OrigPrice             15307
Price                 15308
EnterDate             15308
TransferDate          10923
TransferPrice         15308
TransferInd           15308
TransferBy            15308
CustomerID            15308
ReqPrice              15308
InvoiceNumber         10906
Return                15308
MarkdownDate1          9976
MarkdownDate2          8497
MarkdownDate3          6976
LabelPrinted          15308
MercCut               15307
DownPymt              15308
DownWho                   0
DownDate                  0
ItemCount             15308
Date1                     0
Text1                   395
Num1                  15308
ExpireDate            15308
LayawayGrp                0
PCName                14925
DontPrintTag          15308
ItemNotes                 0
WholesalePrice        15308
BuyersFee             15308
DownPhone                 0
ResaleNow           

In [6]:
df_merchandise['DownPymt'].unique()
# will drop

array([0])

In [7]:
df_merchandise['Num1'].unique()
# will drop

array([0])

In [8]:
df_merchandise['Taxable'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [9]:
df_merchandise['Taxable'].value_counts()
# will drop

Y    14910
N       15
Name: Taxable, dtype: int64

In [10]:
df_merchandise['NeverMD'].unique()

array([nan, 'Y', 'N'], dtype=object)

In [11]:
df_merchandise['NeverMD'].value_counts()
# will drop, unable to find documentation for this column description

N    13159
Y     1766
Name: NeverMD, dtype: int64

In [12]:
df_merchandise['TaxRate'].unique()
# will drop, these values will be worthless

array([nan,  0.])

In [13]:
df_merchandise['CardFee'].unique()
# will drop, these values will be worthless

array([ 0., nan])

In [14]:
# drop columns with no/many missing values and ones that are unnecessary
df_merchandise = df_merchandise.drop(['DownWho','DownDate','Date1','Text1','LayawayGrp','ItemNotes','Num1',
                                      'DownPhone','PreviousOwner','LabelPrinted','PCName','DontPrintTag',
                                      'Taxable','WholesalePrice','AuctionID','CustNo','PhyDate','Auction2',
                                      'DownPymt','ResaleNow','ListOnline','FixSplit','ConCode','NeverMD',
                                      'BuyOut','TaxRate','Reminded','CardFee'], axis=1)

In [15]:
# drop filler row
df_merchandise = df_merchandise.drop([0])
df_merchandise.head()

ProductType ProductColor ProductSize        ProductDescription  \
ProductID                                                                  
-2               None         None        None                      Lamp   
-1               None         None        None  Wrought Iron W. 5 Chairs   
 1               None         None        None             Candle Holder   
 2               None         None        None                     Colum   
 3               None         None        None             Credenza Lamp   

           OrigPrice   Price  EnterDate TransferDate  TransferPrice  \
ProductID                                                             
-2            147.60  147.60 2018-08-18   2018-08-18         147.60   
-1            124.75  124.75 2018-07-03   2018-07-03         124.75   
 1              2.00    2.00 2020-05-20   2020-05-20           2.00   
 2             49.50   49.50 2020-07-18   2020-07-18          49.50   
 3             71.10   71.10 2018-08-25   2018-08-25          71.10   

          TransferInd  ... ReqPrice  InvoiceNumber  Return  MarkdownDate1  \
ProductID              ...                                                  
-2               Sold  ...      0.0         8150.0   False            NaT   
-1               Sold  ...      0.0         7749.0   False            NaT   
 1               Sold  ...      0.0        13211.0   False            NaT   
 2               Sold  ...      0.0        13842.0   False            NaT   
 3               Sold  ...      0.0         8225.0   False            NaT   

           MarkdownDate2 MarkdownDate3 MercCut ItemCount  ExpireDate  \
ProductID                                                              
-2                   NaT           NaT     0.0         0  2018-11-16   
-1                   NaT           NaT     0.5         0  2018-10-01   
 1                   NaT           NaT     0.0         0  2020-08-18   
 2                   NaT           NaT     0.5         0  2020-10-16   
 3                   NaT           NaT     0.0         0  2018-11-23   

           BuyersFee  
ProductID             
-2               0.0  
-1               0.0  
 1               0.0  
 2               0.0  
 3               0.0  

[5 rows x 22 columns]

In [16]:
df_merchandise.dtypes

ProductType                   object
ProductColor                  object
ProductSize                   object
ProductDescription            object
OrigPrice                    float64
Price                        float64
EnterDate             datetime64[ns]
TransferDate          datetime64[ns]
TransferPrice                float64
TransferInd                   object
TransferBy                    object
CustomerID                     int64
ReqPrice                     float64
InvoiceNumber                float64
Return                          bool
MarkdownDate1         datetime64[ns]
MarkdownDate2         datetime64[ns]
MarkdownDate3         datetime64[ns]
MercCut                      float64
ItemCount                      int64
ExpireDate            datetime64[ns]
BuyersFee                    float64
dtype: object

In [17]:
df_merchandise['ProductType'].value_counts()

Accessory          5266
Chair              1194
Decor               655
Table               402
None                384
Pillow              371
Art                 365
Sofa                296
Rug                 293
End Table           228
Lamp                224
Picture             221
Desk                171
Light               155
Bed                 140
Barstool            124
Mirror              113
Coffee Table        111
Dining Table        104
Chest               104
Console              99
Ottoman              93
Cabinet              90
Book                 86
Sidetable            79
Chandelier           77
Bench                65
Nightstand           63
Dresser              60
Clock                54
                   ... 
Set Of 6 Chairs       1
Thomasville Chi       1
Sofa/ottoman          1
PB Bed + Nights       1
Pottery In Stan       1
Bar Stoot             1
Office Unit           1
Wood Bowl             1
Camdle                1
Love Seat             1
Plates          

In [18]:
products = df_merchandise['ProductType'].value_counts()
products[products < 10].shape
# there are 1103 entries where the total count for a given Product Type appears less than 10 times. 
# I will likely exclude these entries from the dataframe when analyzing products

(1103,)

In [19]:
df_merchandise['ProductColor'].value_counts()

None        14779
White          53
Brown          44
Black          29
Ivory          25
Flint          22
Driftwoo       19
Silver         17
Linen          13
Sapphire       12
Wheat          12
Seaglass       12
Gold           11
Blue           11
Red            10
Wood            9
Gray            9
Teal            9
Oatmeal         8
Multi           8
Navy            8
Misc            7
Metal           7
Tommy           6
Clear           6
Yellow          5
Green           5
Bronze          4
Small           4
Rose Oak        3
            ...  
Purple          1
Marble T        1
Verde           1
Gin Patc        1
Gallery         1
White Le        1
Gry/oran        1
Wht Tin         1
Lime            1
Teal/Str        1
Local           1
200.00          1
Birds           1
4 Chairs        1
Blond           1
Blk/Blnd        1
Bernhard        1
Stripe          1
Beige/ta        1
Gray/pur        1
Light Bl        1
Red/Blac        1
2 Leaves        1
Blk Red         1
Yosimite  

In [20]:
products_color = df_merchandise['ProductColor'].value_counts()
products_color[products_color != 'None'].shape

/Users/zachcornelison/anaconda3/lib/python3.7/site-packages/pandas/core/ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


(134,)

Since there's only 134 entires where the Product Color is any description other than "None", I will remove this column from the dataset as it will not be useful for analysis. 

In [21]:
df_merchandise = df_merchandise.drop(['ProductColor'], axis=1)

In [22]:
df_merchandise['ProductSize'].value_counts()

None          14798
Jubilee          94
Park Avenu       39
Queen            18
Small            18
Medium           17
King             14
Large            13
Pomeroy          11
53006225         10
CM3500            8
B446              8
CM1700            8
CM4550            8
CM3200            7
53005996          7
Wood              6
9ft X 4ft         6
22x22             5
Round             3
53005413el        3
Chair             2
4 Pcs             2
V950434           2
98 Inch           2
5X8               2
53050040          2
2311-m            2
53003675el        2
Base              2
              ...  
AR15F01727        1
Triple            1
Bed40k            1
CL19629037        1
30                1
CM3200Q/F         1
Sm2239-3d-        1
22"               1
17x22             1
53005670          1
42x31             1
5140001           1
CLKE144250        1
52005435          1
CM5267            1
Clka1384          1
Bed36q            1
40"               1
D335-425 T        1


In [23]:
products_size = df_merchandise['ProductSize'].value_counts()
products_size[products_size != 'None'].shape

(201,)

I will again drop this column as only 201 of the rows contain useful information regarding Product Size

In [24]:
df_merchandise = df_merchandise.drop(['ProductSize'], axis=1)

In [25]:
df_merchandise['ProductDescription'].value_counts()

Pillow                       611
Lamp                         436
Picture                      347
Art                          315
Chair                        296
Rug                          228
Mirror                       216
Vase                         203
Decor                        202
Chandelier                   145
Barstool                     108
Free                          98
Sofa                          95
Book                          88
Desk                          87
Nightstand                    78
Sidetable                     69
Plate                         68
Finial                        65
Office Chair                  65
Coffeetable                   61
Table                         59
Frame                         56
Pewter Figurine               53
Ottoman                       53
Bookcase                      52
Light Fixture                 51
Painting                      50
Console                       49
Bench                         48
          

want to add product descriptions to product type field if product type field has "none" for the entry and the descrtipion has something relevant to replace with (i.e. sofa, chair, table, rug, etc.).

In [26]:
df_merchandise.head(20)

ProductType        ProductDescription  OrigPrice    Price  \
ProductID                                                               
-2                 None                      Lamp     147.60   147.60   
-1                 None  Wrought Iron W. 5 Chairs     124.75   124.75   
 1                 None             Candle Holder       2.00     2.00   
 2                 None                     Colum      49.50    49.50   
 3                 None             Credenza Lamp      71.10    71.10   
 4                 None                    Pillow     120.00   120.00   
 5                 None                       Rug     269.10   269.10   
 6                 None                   3 Balls       7.50     7.50   
 7         Dining Table       Teak Root/Glass Top    4260.00  4260.00   
 8                 None                       Art       5.50     5.50   
 9                 None                      Lamp      89.00    89.00   
 10                None      Wooden Oklahoma Sign      25.00    25.00   
 11                None                    Candle      20.00    20.00   
 12                None                 Queen Bed     137.50   137.50   
 13                None                       Rug      50.00    50.00   
 14                None                    Mirror      22.00    22.00   
 15                None                 * Unknown      15.12    15.12   
 16                Sofa                C.R. Laine    3600.00  2520.00   
 17                Sofa                C.R. Laine    3600.00  2520.00   
 18                None                Chandelier      27.50    27.50   

           EnterDate TransferDate  TransferPrice TransferInd TransferBy  \
ProductID                                                                 
-2        2018-08-18   2018-08-18         147.60        Sold      Chris   
-1        2018-07-03   2018-07-03         124.75        Sold    Adriana   
 1        2020-05-20   2020-05-20           2.00        Sold       Gina   
 2        2020-07-18   2020-07-18          49.50        Sold       Noel   
 3        2018-08-25   2018-08-25          71.10        Sold    Adriana   
 4        2018-08-30   2018-08-30         120.00        Sold       Noel   
 5        2018-09-28   2018-09-28         269.10        Sold    Adriana   
 6        2018-10-21   2018-10-21           7.50        Sold      Chris   
 7        2015-09-18          NaT           0.00       Store       Noel   
 8        2019-01-02   2019-01-02           5.50        Sold    Adriana   
 9        2019-01-13   2019-01-13          89.00        Sold       Noel   
 10       2019-02-28   2019-02-28          25.00        Sold       Gina   
 11       2019-04-29   2019-04-29          20.00        Sold    Adriana   
 12       2019-08-03   2019-08-03         137.50        Sold       Noel   
 13       2019-09-14   2019-09-14          50.00        Sold       Gina   
 14       2019-10-12   2019-10-12          22.00        Sold       Noel   
 15       2020-01-05   2020-01-05          15.12        Sold       Fred   
 16       2015-09-18          NaT           0.00   Available       Noel   
 17       2015-09-18          NaT           0.00   Available       Noel   
 18       2020-03-04   2020-03-04          27.50        Sold       Gina   

           CustomerID  ReqPrice  InvoiceNumber  Return MarkdownDate1  \
ProductID                                                              
-2                  1       0.0         8150.0   False           NaT   
-1                237       0.0         7749.0   False           NaT   
 1                  1       0.0        13211.0   False           NaT   
 2                237       0.0        13842.0   False           NaT   
 3                  1       0.0         8225.0   False           NaT   
 4                  1       0.0         8262.0   False           NaT   
 5                  1       0.0         8479.0   False           NaT   
 6                  1       0.0         8686.0   False           NaT   
 7                  1       0.0   

In [27]:
df_merchandise.dtypes

ProductType                   object
ProductDescription            object
OrigPrice                    float64
Price                        float64
EnterDate             datetime64[ns]
TransferDate          datetime64[ns]
TransferPrice                float64
TransferInd                   object
TransferBy                    object
CustomerID                     int64
ReqPrice                     float64
InvoiceNumber                float64
Return                          bool
MarkdownDate1         datetime64[ns]
MarkdownDate2         datetime64[ns]
MarkdownDate3         datetime64[ns]
MercCut                      float64
ItemCount                      int64
ExpireDate            datetime64[ns]
BuyersFee                    float64
dtype: object

In [28]:
df_merchandise.InvoiceNumber = df_merchandise.InvoiceNumber.fillna(0.0).astype(int)
df_merchandise.head(3)

ProductType        ProductDescription  OrigPrice   Price  EnterDate  \
ProductID                                                                       
-2               None                      Lamp     147.60  147.60 2018-08-18   
-1               None  Wrought Iron W. 5 Chairs     124.75  124.75 2018-07-03   
 1               None             Candle Holder       2.00    2.00 2020-05-20   

          TransferDate  TransferPrice TransferInd TransferBy  CustomerID  \
ProductID                                                                  
-2          2018-08-18         147.60        Sold      Chris           1   
-1          2018-07-03         124.75        Sold    Adriana         237   
 1          2020-05-20           2.00        Sold       Gina           1   

           ReqPrice  InvoiceNumber  Return MarkdownDate1 MarkdownDate2  \
ProductID                                                                
-2              0.0           8150   False           NaT           NaT   
-1              0.0           7749   False           NaT           NaT   
 1              0.0          13211   False           NaT           NaT   

          MarkdownDate3  MercCut  ItemCount ExpireDate  BuyersFee  
ProductID                                                          
-2                  NaT      0.0          0 2018-11-16        0.0  
-1                  NaT      0.5          0 2018-10-01        0.0  
 1                  NaT      0.0          0 2020-08-18        0.0

In [29]:
df_merchandise.dtypes

ProductType                   object
ProductDescription            object
OrigPrice                    float64
Price                        float64
EnterDate             datetime64[ns]
TransferDate          datetime64[ns]
TransferPrice                float64
TransferInd                   object
TransferBy                    object
CustomerID                     int64
ReqPrice                     float64
InvoiceNumber                  int64
Return                          bool
MarkdownDate1         datetime64[ns]
MarkdownDate2         datetime64[ns]
MarkdownDate3         datetime64[ns]
MercCut                      float64
ItemCount                      int64
ExpireDate            datetime64[ns]
BuyersFee                    float64
dtype: object

In [30]:
# MercCut signifies 0.5 for items consigned (50/50 split), 0.0 for items the store purchased new
df_merchandise.MercCut.value_counts()

0.5    13255
0.0     2052
Name: MercCut, dtype: int64

In [34]:
df_merchandise.ProductType.value_counts().head(20)

Accessory       5266
Chair           1194
Decor            655
Table            402
None             384
Pillow           371
Art              365
Sofa             296
Rug              293
End Table        228
Lamp             224
Picture          221
Desk             171
Light            155
Bed              140
Barstool         124
Mirror           113
Coffee Table     111
Dining Table     104
Chest            104
Name: ProductType, dtype: int64

## Need to clean ProductType column to narrow down the different values

In [37]:
df_merchandise.TransferInd.value_counts()

Sold         10790
Available     3173
Store         1246
Vendor          98
Name: TransferInd, dtype: int64

In [48]:
accessories = df_merchandise[df_merchandise.ProductType == 'Accessory']
accessories.ProductDescription.value_counts().head(25)

Pillow             313
Lamp               269
Picture            226
Art                220
Vase               142
Mirror             129
Decor              118
Rug                 80
Free                72
Chandelier          72
Pewter Figurine     53
Plate               46
Painting            39
Book                37
Finial              36
Frame               35
Floorlamp           34
Light Fixture       34
Bracket             34
Clock               31
Bowl                30
Christmas Decor     29
Greenery            28
Silver Tray         26
Tablelamp           26
Name: ProductDescription, dtype: int64

In [50]:
df_merchandise[df_merchandise.ProductType == 'Pillow'].shape

(371, 20)

In [51]:
df_merchandise[df_merchandise.ProductType == 'Lamp'].shape

(224, 20)

In [52]:
df_merchandise[df_merchandise.ProductType == 'Picture'].shape

(221, 20)

In [53]:
df_merchandise[df_merchandise.ProductType == 'Art'].shape

(365, 20)

In [54]:
df_merchandise[df_merchandise.ProductType == 'Vase'].shape

(45, 20)

In [55]:
df_merchandise[df_merchandise.ProductType == 'Mirror'].shape

(113, 20)

In [56]:
df_merchandise[df_merchandise.ProductType == 'Rug'].shape

(293, 20)

In [57]:
df_merchandise[df_merchandise.ProductType == 'Painting'].shape

(13, 20)

In [ ]:
#df_merchandise = df_merchandise.replace({'ProductType': 'Picture'}, 'Art')